pkg/bff/organization.go

```diff
@@ -197,6 +197,97 @@ func (s *Server) ListOrganizations(c *gin.Context) {
 	c.JSON(http.StatusOK, out)
 }
 
+// DeleteOrganization deletes an organization from the database. The user must have the
+// delete:organizations permission and also be a collaborator in the organization to
+// perform this action.
+//
+// @Summary Delete an organization [delete:organizations]
+// @Description Completely delete an organization, including the registration and collaborators.
+// @Tags organizations
+// @Success 200 {object} api.Reply
+// @Failure 401 {object} api.Reply
+// @Failure 403 {object} api.Reply "User is not a collaborator in the organization"
+// @Failure 404 {object} api.Reply "Organization not found"
+// @Failure 500 {object} api.Reply
+// @Router /organizations/{id} [delete]
+func (s *Server) DeleteOrganization(c *gin.Context) {
+	var (
+		err  error
+		user *management.User
+		org  *models.Organization
+	)
+
+	// Fetch the user from the context
+	if user, err = auth.GetUserInfo(c); err != nil {
+		log.Error().Err(err).Msg("delete organization handler requires user info; expected middleware to return 401")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not identify user to delete organization"))
+		return
+	}
+
+	// Parse the organization ID from the URL
+	orgID := c.Param("orgID")
+
+	// Fetch the organization to be deleted
+	if org, err = s.OrganizationFromID(orgID); err != nil {
+		log.Error().Err(err).Str("org_id", orgID).Msg("could not retrieve organization from database")
+		c.JSON(http.StatusNotFound, api.ErrorResponse("organization not found"))
+		return
+	}
+
+	// The user must be a collaborator in the organization to delete it
+	if org.GetCollaborator(*user.Email) == nil {
+		log.Error().Err(err).Str("org_id", orgID).Str("email", *user.Email).Msg("could not find user in organization collaborators")
+		c.JSON(http.StatusForbidden, api.ErrorResponse("user is not authorized to access this organization"))
+		return
+	}
+
+	// Delete the organization from the database
+	if err = s.db.DeleteOrganization(org.UUID()); err != nil {
+		log.Error().Err(err).Str("org_id", orgID).Msg("could not delete organization from database")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not delete organization from database"))
+		return
+	}
+
+	// Remove the organization from all collaborators so that they don't get an error
+	// when they try to log in. If a user no longer has an organization, they will be
+	// assigned one automatically the next time they log in.
+	for _, collab := range org.GetCollaborators() {
```

Comment:
This is pretty tricky because we're relying on state in Auth0 to decide which organization to attempt to log the user into. If the organization is deleted out from under them, we need to have a way in which they can still login. For most users, what this means is that they will be assigned a new default organization on login. For TSP users who can be assigned to multiple organizations at once, we need to do a bit of extra work to figure out which organization to default them to.

Reply:
If GDS went down in the middle of this operation, could we end up with some collaborators who still have metadata for deleted orgs? If so, are there any places where you think this might cause problems?

Reply:
I think this would mostly cause problems when users try to login; in that case we might have to manually change things in Auth0 to make sure everything is in sync.